src/esp/gfx/CubeMap.cpp

```diff
@@ -430,20 +533,83 @@ void CubeMap::loadTexture(TextureType type,
         CORRADE_INTERNAL_ASSERT_UNREACHABLE();
         break;
 
+      case TextureType::VarianceShadowMap:
+        // TODO: vsm texture
+        CORRADE_INTERNAL_ASSERT_UNREACHABLE();
+        break;
+
       case TextureType::Count:
         CORRADE_INTERNAL_ASSERT_UNREACHABLE();
         break;
     }  // switch
     LOG(INFO) << "Loaded image " << iFace << " from " << filename;
   }
-  // Color texture ONLY, NOT for depth
-  if ((flags_ & Flag::BuildMipmap) && (flags_ & Flag::ColorTexture)) {
+  // sanity check is within the function
+  generateMipmap(type);
+}
+
+void CubeMap::generateMipmap(TextureType type) {
+  CORRADE_INTERNAL_ASSERT(type == TextureType::Color ||
+                          type == TextureType::VarianceShadowMap);
+  if (type == TextureType::Color) {
+    CORRADE_INTERNAL_ASSERT(flags_ & Flag::ColorTexture);
+  } else if (type == TextureType::VarianceShadowMap) {
+    CORRADE_INTERNAL_ASSERT(flags_ & Flag::VarianceShadowMapTexture);
+  }
+  if ((flags_ & Flag::AutoBuildMipmap) ||
+      (flags_ & Flag::ManuallyBuildMipmap)) {
+    Mn::GL::CubeMapTexture& tex = texture(type);
     tex.generateMipmap();
   }
 }
 
+void CubeMap::copySubImage(unsigned int cubeSideIndex,
+                           TextureType type,
+                           Magnum::GL::Texture2D& texture,
+                           unsigned int mipLevel) {
+  CORRADE_INTERNAL_ASSERT(cubeSideIndex < 6);
+
+  if (mipLevel > 0) {
+    mipLevelSanityCheck("CubeMap::CopyToTexture2D():", flags_, mipLevel,
+                        mipmapLevels_);
+  }
+
+  int size = imageSize_ / pow(2, mipLevel);
+
+#ifndef MAGNUM_TARGET_WEBGL
+  CORRADE_ASSERT(texture.imageSize(0) == Mn::Vector2i(size, size),
```

Comment:
FYI, the single-argument constructor does exactly this:

```suggestion
  CORRADE_ASSERT(texture.imageSize(0) == Mn::Vector2i{size},
```